overlord/ifacestate/handlers.go

```diff
@@ -507,6 +507,26 @@ func (m *InterfaceManager) doDisconnect(task *state.Task, _ *tomb.Tomb) error {
 	return nil
 }
 
+// timeout for shared content retry
+var contentLinkRetryTimeout = 30 * time.Second
+
+// defaultContentProviders returns a dict of the default-providers for the
+// content plugs for the given snapName
+func (m *InterfaceManager) defaultContentProviders(snapName string) map[string]bool {
+	plugs := m.repo.Plugs(snapName)
+	defaultProviders := make(map[string]bool, len(plugs))
+	for _, plug := range plugs {
+		if plug.Interface == "content" {
+			if s, _ := plug.Attrs["content"].(string); s != "" {
+				if dprovider, _ := plug.Attrs["default-provider"].(string); dprovider != "" {
+					defaultProviders[dprovider] = true
+				}
+			}
+		}
+	}
+	return defaultProviders
```

Comment:
What happens if a snap declares self as a default-provider (developer's mistake)? Should we actively prevent that here?

Reply:
I think with the last code is a noop, on the other end is strange.
Not sure though if the same applies to bases and if we have a check for that.

Reply:
I added a check that base/os snaps cannot have a base themself (this will error now in validate.go). A default-provider to itself would be strange but would work, the snapstate code will see that the snap is already in the change as an install so it will do nothing. And the autoconnect part will just check if the snap was linked (which it was) so that is fine too.